feat: fluent-bit example (#5035)
talos-systems/talos: website/content/docs/v0.15/Guides/logging.md

This ensures filebeat uses the host network, and listens on `127.0.0.1:12345`
(UDP) on every machine, which can then be specified as a logging endpoint in
the machine configuration.

### Fluent-bit example

> This example shows to us, how to collect Talos and pod logs in the cluster.

Reviewer: remove this line?

> First, we'll create a value file for fluentd-bit helm-chart.

Reviewer: ```suggestion
First, we'll create a value file for the `fluentd-bit` Helm chart.
```

      K8S-Logging.Exclude On
      Annotations         Off
      Labels              On

    [FILTER]
      Name          modify
      Match         kubernetes.*
      Add           source kubernetes
      Remove        logtag

  customParsers: |
    [PARSER]
      Name          audit
      Format        json
      Time_Key      requestReceivedTimestamp
      Time_Format   %Y-%m-%dT%H:%M:%S.%L%z

    [PARSER]
      Name          containerd
      Format        regex
      Regex         ^(?<time>[^ ]+) (?<stream>stdout|stderr) (?<logtag>[^ ]*) (?<log>.*)$
      Time_Key      time
      Time_Format   %Y-%m-%dT%H:%M:%S.%L%z

  outputs: |
    [OUTPUT]
      Name    stdout
      Alias   stdout
      Match   *
      Format  json_lines

daemonSetVolumes:
  - name: varlog
    hostPath:
      path: /var/log

daemonSetVolumeMounts:
  - name: varlog
    mountPath: /var/log

tolerations:
  - operator: Exists
    effect: NoSchedule
```

Next, we will add the helm repo for FluentBit, and deploy it to the cluster.

```shell
$ helm repo add fluent https://fluent.github.io/helm-charts
$ helm upgrade -i --namespace=kube-system -f fluentd-bit.yaml fluent-bit fluent/fluent-bit
```

> Then, we need to know the service IP.

Reviewer: ```suggestion
Now  we need to find the service IP.
```

```shell

> $ kube -n kube-system get svc -l app.kubernetes.io/name=fluent-bit

Reviewer: ```suggestion
$ kubectl -n kube-system get svc -l app.kubernetes.io/name=fluent-bit
```

NAME         TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)             AGE
fluent-bit   ClusterIP   10.200.0.138   <none>        2020/TCP,5170/TCP   108m
```

> Finally, we will change talos log destination by command ```talosctl edit mc --immediate```.

Reviewer: ```suggestion
Finally, we will change talos log destination by command ```talosctl edit mc```.
```

Talos now automatically applies changes in immediate mode when possible.

> Finally, we will change talos log destination by command ```talosctl edit mc --immediate```.

Reviewer: ```suggestion
Finally, we will change talos log destination with the command ```talosctl edit mc --immediate```.
```

      Time_Key      time
      Time_Format   %Y-%m-%dT%H:%M:%S.%L%z

  outputs: |
    [OUTPUT]
      Name    stdout
      Alias   stdout
      Match   *
      Format  json_lines

daemonSetVolumes:
  - name: varlog
    hostPath:
      path: /var/log

daemonSetVolumeMounts:
  - name: varlog
    mountPath: /var/log

tolerations:
  - operator: Exists
    effect: NoSchedule
```

Next, we will add the helm repo for FluentBit, and deploy it to the cluster.

```shell
$ helm repo add fluent https://fluent.github.io/helm-charts
$ helm upgrade -i --namespace=kube-system -f fluentd-bit.yaml fluent-bit fluent/fluent-bit
```

Now  we need to find the service IP.

```shell
$ kubectl -n kube-system get svc -l app.kubernetes.io/name=fluent-bit

NAME         TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)             AGE
fluent-bit   ClusterIP   10.200.0.138   <none>        2020/TCP,5170/TCP   108m
```

Finally, we will change talos log destination with the command ```talosctl edit mc```.

```yaml
machine:
  logging:
    destinations:
      - endpoint: "tcp://10.200.0.138:5170"
        format: "json_lines"
```

> This example well tested with cilium, and has to works with iptables/ipvs based cni plugins too.

Reviewer: ```suggestion
This example configuration was well tested with Cilium CNI, and it should work with iptables/ipvs based CNI plugins too.
```
